[WIP] LMNN distance learning

manish7294 opened the pull request:
Currently, I am pushing a rough implementation of LMNN. A lot of things will be updated within this PR including algorithms, tests, comments, etc.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+  math::ClearAlias(labels);
+
+  dataset = std::move(newDataset);
+  labels = std::move(newLabels);
```
Each time the dataset is shuffled, I believe we have to recalculate `TargetNeighbors()`.

rcurtin replied:
Or, on second thought, it's not necessary to recalculate the target neighbors, but we do need to apply the same shuffling to the indices.  I guess an easy way to do it for testing is just to call `TargetNeighbors()` again with the shuffled data (but that is very slow and we shouldn't use it as a long-term solution).

manish7294 replied:
I read shuffleData() code and I think instead of directly calling shuffleData() we can use its internal code ( as it will give us the shuffle ordering ) and just re-map target neighbors to ordering indices. What do you think?

manish7294 replied:
I think using variadic won't work ---- here in case of target neighbors, all the target neighbors will have different indexes (resulting in a change of all row elements too). So, just changing order of target neighbors matrix cols according to shuffling order won't result in correct target neighbors matrix.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+    }
+  }
+
+  return cost;
```
This looks correct to me.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+    }
+  }
+
+  return cost;
```
This also looks correct to me.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+}
+
+//! Calculate outer product of vector.
+arma::mat OuterProduct(arma::vec& diff)
```
I think it's unnecessary to have this function (I think it's just fine to write it in-place), but if you do want to include it, I'd suggest inlining it.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+      // Calculate gradient due to target neighbors.
+      arma::vec cij = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+      gradient += (1 - regularization) * OuterProduct(cij);
+    }
```
I think that you could also move the `gradient += (1 - regularization) * OuterProduct(cij)` outside the loop, and just take the sum of all `cij`s inside the loop.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+//! Compute gradient over whole dataset.
+template<typename MetricType>
+void LMNNFunction<MetricType>::Gradient(const arma::mat& coordinates,
+                                        arma::mat& gradient)
```
This assumes the same impostors as `Evaluate()` (or, I guess, the real assumption is that `Gradient()` will always be called after `Evaluate()` with the same `coordinates`).  I think that is true in most cases, but if you are having trouble getting the algorithm to converge correctly, you could always try manually recalculating impostors here.  In addition, if you implement an `EvaluateWithGradient()` function, an optimizer can make use of that and you can save on the calculation of impostors there.  (See http://mlpack.org/docs/mlpack-git/doxygen/optimizertutorial.html, section "The FunctionType API", if you haven't already)

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+  arma::cx_vec eigval;
+  arma::cx_mat eigvec;
+
+  arma::eig_gen(eigval, eigvec, iterate);
```
`iterate` has to be symmetric (since we are optimizing M), so you can use `eig_sym()` here which will make your life a bit easier.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+template<typename DecomposableFunctionType>
+double StandardSGD::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
```
I think that this will work, but I'm not sure how `Projection()` will work for L_BFGS.  I suspect that maybe you could use a custom update policy for SGD to get the projection in there (instead of this overload).  I'm not sure of the best way to do it for L-BFGS, I'd have to think about it a bit...

rcurtin replied:
Hmm, so it depends on whether or not we plan to go with this approach in the long term.  In the short term, for testing, we can do something *really* ugly and at each `Evaluate()` call, cast the `const arma::mat& coordinates` to an `arma::mat&` then project it.  That can work for testing but it is, in my view, far too much of a hack to actually commit long-term.

Alternately, I could see overloading a few optimizers to handle the projection step and then using a `static_assert()` to ensure that an overloaded optimizer is used, or, modifying all of the optimizers to have a `PostStepPolicy` or something like this that allows some action to be taken after a step.  (That last idea would be a lot of work for just one `FunctionType` though.)

manish7294 commented:
Sure, I will add some tests shortly.

rcurtin commented:
Sounds good---once the tests are there, we can know whether or not LMNN is working right.  We can try to reproduce the results in the paper, and see how long it takes to run.  Then we could switch out the M matrix for L^T L and this means there's no need for a projection step or any of the ugly hacks necessary for that, and then we can see if that's faster.  Let me know if you think that's a reasonable plan, and if you already had something else in mind we could do that instead. :)

manish7294 commented:
That sounds totally well put plan. I am more than happy to go with it :)

manish7294 commented:
@rcurtin I made a quick comparison of shogun's LMNN with our's current implementation on dataset given in this [notebook](http://shogun-toolbox.org/static/notebook/current/LMNN.html) with k = 1.

Here is a graph for original dataset - [link](https://pasteboard.co/HmDq0VB.png) 
Here is result by shogun - [link](https://pasteboard.co/HmDqdP6.png)
Here is result by current implementation using L-BFGS optimizer - [link](https://pasteboard.co/HmDs7nK.png)

rcurtin commented on src/mlpack/tests/lmnn_test.cpp:
```diff
@@ -0,0 +1,352 @@
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6);
+
+  double initObj = lmnnfn.Evaluate(arma::eye<arma::mat>(2, 2));
+  double finalObj = lmnnfn.Evaluate(outputMatrix);
```
Hmm, this checks that the objective function is better, which is correlated to the result we want, but what we'd really like to see is that the nearest neighbor of each point now has the same label.  So maybe we should check that explicitly?

manish7294 replied:
I think, "final nearest neighbors having same labels" shouldn't be necessary condition. As that would be the ideal case. What do you think? 
Maybe we can do that check for this dataset explicitly.

rcurtin replied:
Oops, didn't respond to this one, sorry about that.  I think maybe it is worth checking that KNN accuracy improves for this very simple dataset, but you are right, in general that may not always be the case.

manish7294 replied:
I made a run on this dataset. We are getting 100% accuracy with this.

rcurtin replied:
Sounds good, we should write that into the test either by manual distance calculations or using the KNN object.

manish7294 commented:
@rcurtin One thing is still bugging me. Currently we are getting good enough results with L-BFGS  but with sgd the results are not good. Do you have some idea as what could be happening?

rcurtin commented:
How many points are you using for testing?  If the dataset is a tiny 6-point example dataset, then the SGD gradients for each mini-batch (or for each point, if the batch size is 1) are likely to point in opposite directions, and the optimizer may have a harder time converging.  I'd suggest making a larger dataset (hundreds or thousands of points perhaps) and trying that; I think maybe it will work better.

Alternately, the step size may just need to be tuned.  You could print the objective at every iteration to see what is going on too---if the step size is too large, it may diverge.

manish7294 commented:
Right, that may be the cause. As with a dataset (iris)  just greater than 6 points, the results were better but not better than L-BFGS. I will report it after checking once more.

manish7294 commented:
I took a look at objective after every iteration and the final objective, to my surprise sgd's final objective is always coming to be the highest among all the values. Don't know what's going on?

rcurtin commented:
Looks like the learning rate is too large and it is bouncing around minima.  Try reducing the learning rate significantly... if you reduce it too much, it will be very slow to reduce the objective, but if you don't reduce it enough, it'll decrease for a bunch of steps and then may explode back to large values.  This is a disadvantage of SGD, sometimes it needs to be tuned. :)

manish7294 commented:
@rcurtin you're right, here's a result with same above dataset using sgd: [link](https://pasteboard.co/HmZABXv.png)

manish7294 commented:
I think somehow as per figure, target neighbors are also being pushed away.

rcurtin commented:
(sorry, I was away over the weekend and it was a 3-day weekend here)

The figure looks ok, but not perfect.  Is that figure generated after the step size is tuned, or is that from the SGD output you gave earlier?

manish7294 commented:
The figure is generated after tuning.

rcurtin commented:
You can also compare the objective function value with the one generated by L-BFGS.

manish7294 commented:
Ya, final objective value is around 2.7 as compare to starting 3.6 which is close to L-BFGS one.

manish7294 commented:
Sure,  I will tell you the results shortly.

rcurtin commented:
Keep in mind also, SGD has a random component, so you should probably run the test a handful of times for each dataset to ensure that it's converging to somewhere close to the same.  (Also the step size may need a little bit of tuning, possibly.)

manish7294 commented:
I think there is still some kind of problem going on here. As I increase the step size the objective value after some iterations continuously keep on increasing, causing final objective to be very large. As per the results ---- this increase is scaling datasets to large values.
Here are some examples - 
at step_size 0.01 [link](https://pasteboard.co/HnwXpmM.png)
at step_size 0.1 [link](https://pasteboard.co/HnwYuLw.png)
And setting different seeds is not affecting final objective at all. I am not getting what is happening here. 
And one more thing, I think inclusion/exclusion of projection step is not causing any change.

manish7294 commented:
I have tried running the current implementation on full MNIST dataset, but even after hours of execution --- several times, I end up losing ssh connection :(
The calculation of impostors after every iteration is just too costly.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,395 @@
+  // Hack
+  arma::mat& iterate = const_cast<arma::mat&>(coordinates);
+  // Apply projection.
+  if (arma::accu(iterate == initialPoint) != initialPoint.n_cols)
```
Should this be comparing against `initialPoint.n_elem`?

manish7294 replied:
Right, I came across & realize that this morning, but haven't pushed the change to the remote.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,395 @@
+        diff = dataset.col(i) - dataset.col(impostors(l, i));
+        cil -= diff * arma::trans(diff);
+      }
+    }
```
When I reviewed this the first time I thought it was okay, but I realized I forgot about the `std::max(0.0, 1 + eval)`, and I don't think that `max()` is accounted for here.  I *think* that we can say that if the constraint is satisfied, the gradient will be 0 and we can ignore it.  I guess the condition would be `eval < -1`.  I am not 100% sure on the derivation there, but I think it is at least the right direction.

manish7294 replied:
Hmm, This looks like a valid point. But I don't think the authors have considered this. I think we can try imposing this condition. 
>  I guess the condition would be eval < -1

This condition seems reasonable. So, it means we will be calculating eval once again, here in gradient. Right?

rcurtin replied:
Ack, sorry, I read the `[ ... ]_+` from the evaluation bit.  Sorry for the error on my part.

But note that the gradient computation (the equation below (18)... why isn't it numbered?) does not use every possible tuple.  Instead it only uses some:

"Let us define a set of triples N_t, such that (i, j, l) ∈ N_t if and only if the indices (i, j, l) trigger the hinge loss in the second part of Eq. (18)."

So I do think we have to compute `eval` again in order to find the set of triples N_t that we need to use for the gradient.

manish7294 replied:
> Let us define a set of triples N_t, such that (i, j, l) ∈ N_t if and only if the indices (i, j, l) trigger the hinge loss in the second part of Eq. (18).

Right, this line contains the thing. Okay, I will make the required change and will look into the results.

manish7294 replied:
@rcurtin, this was the exact problem. Now, I ran quite a number of tests on multiple dataset and everything looks promising. :) 
Here is a result: [link](https://pasteboard.co/HnB3o5x.png)

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,395 @@
+
+  arma::eig_sym(eigval, eigvec, iterate);
+
+  arma::uvec ind = arma::find(eigval > 0);
```
You could check to see if `Projection()` is ever doing anything with something like this:

```
if (ind.n_elem != eigval.n_elem)
  Log::Warn << "Removed " << (eigval.n_elem - ind.n_elem) << " eigenvectors!" << std::endl;
```

manish7294 commented:
I ran quite a number of tests and everything seems pretty good. Each run converged successfully and reduced the objective to much lower value. Apart from the standard datasets, I ran the LMNN on your NCA toy dataset too and results were same as expected (the similar label points came to a distance of 0.03 from 1, whereas impostors went beyond 0.8 from an initial distance of 0.2 leaving objective very close to 0).

Here is a result using iris dataset - 
Initial - [link](https://pasteboard.co/HnG37XU.png)
Final - [link](https://pasteboard.co/HnG3jMu.png)
(I made a graph between axis 1 - 2 & 3 - 4 -- other combinations too have the similar kind of improvement )

manish7294 commented:
Benchmarks on Iris:
SGD:
K = 5 -> Accuracy Initial: 96.6667, Accuracy Final: 97.3333, total_time: 34.330329s
K = 10 -> Accuracy Initial: 97.3333, Accuracy Final: 98.6667, total_time: 63.079647s
K = 15 -> Accuracy Initial: 97.3333, Accuracy Final: 99.3333, total_time: 97.608639s

L-BFGS:
K = 5 -> Accuracy Initial: 96.6667, Accuracy Final: 96.6667, total_time: 1.180102s
K = 10 -> Accuracy Initial: 97.3333, Accuracy Final: 98.0000, total_time: 2.660565s
K = 15 -> Accuracy Initial: 97.3333, Accuracy Final: 99.3333, total_time: 5.215524s

rcurtin commented:
Looks great!  I agree that it works.  It looks like we are a few weeks ahead of your proposal's schedule, which is great.  So I think that gives us some good time for optimization.  I'll go through the code and leave a few comments in places where I think the code could be accelerated.

Another useful idea is to use a profiler like gprof or perf (or something else) to see what parts of the program are taking the longest.  You can also use the `Timer` class to do some simpler profiling, i.e.,

```
Timer::Start("my_loop");
for (...)
{
  // lots of computation or whatever
}
Timer::Stop("my_loop");
```

and if you've written some `lmnn_main.cpp` or something, when you run that with `-v` it will print the total time of each of those timers.  (I guess that's what you're using for your simulations in the previous comment.)

manish7294 commented:
Right, I am using verbose here and It is a providing a great log of timing separation of each part.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,415 @@
+  transformedDataset = coordinates * dataset;
+
+  Constraints constraint(transformedDataset, labels, k);
+  constraint.Impostors(impostors);
```
Note that we only need some of the impostors---specifically those for points with indices between `begin` and `begin + batchSize`.  So you could reduce the size of the KNN query to save some time.

manish7294 replied:
This is nice speedup :)

rcurtin commented on src/mlpack/methods/lmnn/constraints_impl.hpp:
```diff
@@ -0,0 +1,148 @@
+    // Perform KNN search with differently labeled points as reference
+    // set and  same class points as query set.
+    knn.Train(dataset.cols(indexDiff));
+    knn.Search(dataset.cols(indexSame), k, neighbors, distances);
```
When we call `knn.Train()`, this builds a tree on the reference points.  That tree building process can take a long time, depending on the data.  (For instance, with MNIST it can take a while.)  However, KNN allows the use of a custom distance metric, and mlpack has the `MahalanobisDistance` metric in `src/mlpack/core/metrics/`.  So, it is almost possible to do something a little like this:

```
MahalanobisDistance md(dimensionality);
NeighborSearch<NearestNeighborSort, MahalanobisDistance> knn(referenceSet);
for (size_t i = 0; i < iterations; ++i)
{
  arma::mat parameters = GetParametersSomehow(); // just an example...
  knn.Metric().Covariance() = parameters;

  knn.Search(querySet, k, neighbors, distances);

  ...
}
```

Of course that is not the exact structure of the code that you have, but hopefully you see where I am going with it, which is that you can build the kd-tree once, then change the MahalanobisDistance covariance matrix, and get new results.

But there are some gotchas:

 * The `BinarySpaceTree<>` class also has a bounding structure.  By default `HRectBound` is used, but that will not work with `MahalanobisDistance`.  We would need to write a `MahalanobisHRectBound` class or something like this that used the distance matrix from the Mahalanobis distance.

 * As the Mahalanobis distance matrix gets further and further away from the identity matrix, the nearest neighbor search will become slower and slower as the bounds of the hyperrectangles in the tree change.  So at some point it becomes worthwhile to rebuild the tree.

 * You can also build the query tree only once to save some time.  By default each call to `Search()` will build a tree on the query points too, but you could pass in a pre-built tree.

This is kind of an in-depth optimization and would take a little time and implementation, but it could help speed things up.

manish7294 replied:
This could result in good optimization. Will have to think more about the code structure. I will update you as soon as I get something.

manish7294 replied:
Let's say we want to make this optimization to impostors part only (as we generally keep target neighbors constants except when shuffle is used, here also calling TargetNeighbors() can be avoided). So, if we want to employ this change we will need to store at least `uniqueLabels.n_elem` number of trees. Do you think this is efficient and do we have a way to do this?

rcurtin replied:
Each tree should have size somewhat less than 2x the dataset it is built on.  I think it's not likely that a user will manage to use a dataset that is so large that memory becomes an issue, since the optimization is sufficiently slow right now that I don't think we could do 100k points in any reasonable amount of time.  If we do accelerate it to that fast then maybe memory will become an issue then, but for now I wouldn't worry about it.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,415 @@
+
+  // Re-calculate impostors on transformed dataset.
+  Constraints constraint(transformedDataset, labels, k);
+  constraint.Impostors(impostors);
```
So, here we always recompute impostors, but we may not need to.  Let's say that `k = 3`, so for some point `p_i` and coordinates `M_t` we have `d_1(p_i)`, `d_2(p_i)`, and `d_3(p_i)`, which are the distances to the nearest impostor, second nearest impostor, and third nearest impostor.  Suppose that we also computed `d_4(p_i)`, the distance to the fourth nearest impostor.  When we take a step and get `M_{t + 1}`, we should be able to produce a lower bound on `d_4(p_i)` and an upper bound on `d_1(p_i)`, `d_2(p_i)`, and `d_3(p_i)`.  If that lower bound is still greater than any of the upper bounds, then the impostors cannot change and there is no need to recompute the impostors with a nearest neighbor search.

However I haven't yet had a chance to compute what those bounds actually would be.  I am thinking that I can probably express some bound with respect to the matrix norm || M_t - M_{t + 1} ||.  If you have a chance, do you want to give the bound a shot?

manish7294 replied:
If I understand this correctly, then we want to measure the change in impostors distance cause by each iteration and if it is still lower than lower bound (which we get by the distance of 
'k+1'th impostors ) then we don't need to recalculate impostors as they will still be the same. Please correct me if I am not correct here. 
If we can gain speedups, then I sure would like to work on this.

rcurtin replied:
That's correct.  The only thing is this is one of the more difficult speedups, so it's probably better to start with the simpler ideas and see how much they help.

manish7294 replied:
I spent some time with this and got a rough idea, I will be writing it down below. Please, Let me know if it sounds reasonable.

At every iteration the change in position of any data point is a measure of ( M_{t + 1} - M_{t} ) or more specifically  || ( M_{t + 1} - M_{t} ) *  p_{i} || . 
So what we can do is, keep an oldCoordinates and an oldImpostors (upto k + 1 impostors) matrices and then iterate over each impostors column (impostors corresponding to each data point, so total N columns) --- here within this iteration we will compare || ( M_{t + 1} - M_{t} ) *  p_{i} || ^ 2 for each impostor from i = 1 to i = k with  || M_{t} * ( p_{k + 1} - p_{i}) || ^ 2 and if the first term exceeds 2nd term for any of the i 's, then we will store the index of this data point in a Index vector ---- After performing the same operation for all the data points, we will have final Index vector (it will have indexes of those data points whose impostors are changed) and then we can calculate the impostors for just these points (Index vector may be a null vector too).

rcurtin replied:
Right, I think that that will work (I am taking L to be the same as M in your writing).  I did some derivation on the rest of the open whiteboard, attached below.  I want to write it up more formally but maybe it will give you an idea.  Basically I wanted to prove, if we know || L_{t + 1} - L_t ||, can we use that to bound which points could change impostors?  I have most of the result in the image but I don't know how legible it is.  I'll return to the idea tomorrow but I think I will end up with something roughly similar to the idea you have here.

![img_20180604_164437](https://user-images.githubusercontent.com/1845039/40940847-12c13e5c-6817-11e8-8747-89e7d53595a1.jpg)

If || L_{t + 1} - L_t || is sufficiently small, then we could know that impostors cannot change for any point at all, and thus we could avoid recalculation.

Alternately: when using SGD, there is "noise" in the gradient anyway by virtue of only calculating the gradient with respect to a few points at a time.  So if we were to say "we only recalculate impostors once every N passes through the data", or, "we only recalculate impostors when || L_{t + 1} - L_t || > a for some value a", that might still converge just fine.

I'll try and write the notes up into something more coherent tomorrow.  Sorry if they are not fully legible.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,415 @@
+                      metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(impostors(l, i)));
+
+        cost += regularization * std::max(0.0, 1 + eval);
```
We can do a slight shortcut here (and in other similar places)---if `eval` is less than -1 for `l`, then `eval` will also be less than -1 for `l + 1` so we can break early.

Similarly, if we iterated backwards over `j`, then if `eval` is less than -1 for `l`, then `eval` will also be less than -1 for `j - 1`, etc., so we can break early in that case too.

These two ideas can also be combined.

manish7294 replied:
Yeah! This could help.

manish7294 commented:
> Note that we only need some of the impostors---specifically those for points with indices between begin and begin + batchSize. So you could reduce the size of the KNN query to save some time.

> We can do a slight shortcut here (and in other similar places)---if eval is less than -1 for l, then eval will also be less than -1 for l + 1 so we can break early.
Similarly, if we iterated backwards over j, then if eval is less than -1 for l, then eval will also be less than -1 for j - 1, etc., so we can break early in that case too.

Currently, I have completed the above two optimizations and results are awesome. The running time approximately reduced to half as k increases on iris.

Earlier:
K = 5 - total_time: 34.330329s
K = 10 - total_time: 63.079647s
K = 15 - total_time: 97.608639s

Now:
K = 5 - total_time: 22.663736s
K = 10 - total_time: 36.749581s
K = 15 - total_time: 48.495706s

rcurtin commented:
Ah, by the way, are you configuring your build without debugging symbols and with optimization?  (i.e. `cmake -DDEBUG=OFF`) That can give large speedups if you haven't done it already.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,459 @@
+  double cost = 0;
+
+  // Apply metric over dataset.
+  transformedDataset = coordinates * dataset;
```
I didn't realize it when I reviewed, but it looks like every use of `coordinates` implies that the implementation here actually is optimizing `L` for `M = L^T L`.  That means two things:

 * No projection step is necessary---for any `L`, `M` is guaranteed to be positive semidefinite.

 * It should be possible to allow the user to specify low-rank `L`: that is, instead of being `d x d`, `L` can have size `r x d` for some user-specified `r <= d`.  Depending on the dataset, some smaller values of `r` may give degraded performance.  But, small `r` will generally both accelerate the KNN search in the transformed space since it is lower-dimensional, and also accelerate the optimization process since the matrices are smaller.

In general for your optimization efforts I would say we should add this but don't get hung up if it is not giving perfectly good performance.

manish7294 replied:
>  it looks like every use of coordinates implies that the implementation here actually is optimizing L for M = L^T L.

Can you please explain some more, I can't seem to get it.

rcurtin replied:
So, in the paper, we are trying to find some Mahalanobis distance matrix `M`.  This matrix `M` improves classification performance when we use the distance function `d_M(x, y) = (x - y)^T M (x - y)`.

But note that if we stretch the dataset (as is done in the code) by some matrix `L` as `X' = L*X`, this would give the distance function `d'(x, y) = (Lx - Ly)^T (Lx - Ly) = (x - y)^T L^T L (x - y)`.  Hence, `L` must be such that `M = L^T L`.

So, if we want to optimize over `M`, then the code should be, e.g., `transformedDataset = arma::chol(coordinates) * dataset` (another decomposition could work too, maybe LU is faster?).  But in the end, I'd like to optimize over `L`, because then we can also make it low-rank which could give additional speedup (and was the original idea behind LMNN+LRSDP anyway).

Does that make sense?  Let me know if I can clarify further.

manish7294 replied:
> if we stretch the dataset (as is done in the code) by some matrix L as X' = L*X, this would give the distance function d'(x, y) = (Lx - Ly)^T (Lx - Ly) = (x - y)^T L^T L (x - y). Hence, L must be such that M = L^T L.

Thanks, this helps a lot. I was missing the point that we are actually learning L here, it became clear as you explained.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,459 @@
+        }
+
+        arma::vec diff = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+        cil += diff * arma::trans(diff);
```
As a follow-up on the low-rank idea, I think this bit will need to be re-derived.  I'll take a look into that after I look into the bounds for the impostors that you wrote up.

manish7294 replied:
Now, if we partially derivate the gradient w.r.t L (instead of M) we may just get an extra factor of multiple of L (coordinates). As far I could see all other things should remain same. Though, I am not sure about this.

rcurtin replied:
I think you are right, but I haven't double-checked yet.

rcurtin replied:
What I computed for the quantity df/dL:

```
df/dL = (1 - mu) * \sum_{i, j} 2 L (x_i - x_j) (x_i - x_j)^T + \sum_{i, j, l \in N} (2 L (x_i - x_j) (x_i - x_j)^T - 2 L (x_i - x_l) (x_i - x_l)^T).
```

Full whiteboard derivation attached:
![img_20180604_160755](https://user-images.githubusercontent.com/1845039/40939057-936d701c-6811-11e8-99a7-257629192f80.jpg)

rcurtin replied:
Note that the quantity is actually a little different than before, we can't just use `transformedDataset` to compute the gradient.  Actually we have to compute the gradient in the original space and multiply `L`.  That also means we can precompute the quantities `sum_{i, j} (x_i - x_j)(x_i - x_j)^T` and probably the second term but there may be a little more trickiness there.

manish7294 replied:
Right, I think the formulation you made is absolutely correct, Earlier I also got the same exact formulation. I think we can rely on this.

manish7294 replied:
>  we can precompute the quantities sum_{i, j} (x_i - x_j)(x_i - x_j)^T and probably the second term but there may be a little more trickiness there.

We can surely precalculate the first term here but as you already said, precomputing second part is not so straightforward and I think it might not be possible even. 

I tried the above formulation and results are fine but I am facing some obstacles:
1. For a comparatively large value of step_size objective keep on increasing to inf.
2. At some value of step_size the objective decreases to minima and they again slowly starts to increase leaving final objective greater than earlier minima.
3. This shouldn't be happening but at some value of k, final accuracy is less than initial ones.

rcurtin replied:
Can you push the implementation?  One note is in the derivation above, the second term only applies to those triplets i,j,k where `1 + (x_i - x_j)^T L^T L (x_i - x_j) - (x_i - x_l)^T L^T L (x_i - x_l))` is greater than 0.

manish7294 replied:
I just pushed the changes regarding lmnn_function_impl.hpp, please have a look at them.

rcurtin replied:
I took a look and I added some numeric gradient checker tests like we do for the ANN layers, and everything seemed fine:

```
diff --git a/src/mlpack/tests/lmnn_test.cpp b/src/mlpack/tests/lmnn_test.cpp
index 43020988a..f021a0595 100644
--- a/src/mlpack/tests/lmnn_test.cpp
+++ b/src/mlpack/tests/lmnn_test.cpp
@@ -370,4 +370,113 @@ BOOST_AUTO_TEST_CASE(LMNNLBFGSSimpleDataset)
   BOOST_REQUIRE_LT(finalObj, initObj);
 }
 
+// Simple numerical gradient checker.
+template<class FunctionType>
+double CheckGradient(FunctionType& function,
+                     arma::mat& coordinates,
+                     const double eps = 1e-7)
+{
+  // Get gradients for the current parameters.
+  arma::mat orgGradient, gradient, estGradient;
+  function.Gradient(coordinates, orgGradient);
+
+  estGradient = arma::zeros(orgGradient.n_rows, orgGradient.n_cols);
+
+  // Compute numeric approximations to gradient.
+  for (size_t i = 0; i < orgGradient.n_elem; ++i)
+  {
+    double tmp = coordinates(i);
+
+    // Perturb parameter with a positive constant and get costs.
+    coordinates(i) += eps;
+    double costPlus = function.Evaluate(coordinates);
+
+    // Perturb parameter with a negative constant and get costs.
+    coordinates(i) -= (2 * eps);
+    double costMinus = function.Evaluate(coordinates);
+
+    // Restore the parameter value.
+    coordinates(i) = tmp;
+
+    // Compute numerical gradients using the costs calculated above.
+    estGradient(i) = (costPlus - costMinus) / (2 * eps);
+  }
+
+  // Estimate error of gradient.
+  return arma::norm(orgGradient - estGradient) /
+      arma::norm(orgGradient + estGradient);
+}
+
+BOOST_AUTO_TEST_CASE(LMNNFunctionGradientTest)
+{
+  // Useful but simple dataset with six points and two classes.
+  arma::mat dataset        = "-0.1 -0.1 -0.1  0.1  0.1  0.1;"
+                             " 1.0  0.0 -1.0  1.0  0.0 -1.0 ";
+  arma::Row<size_t> labels = " 0    0    0    1    1    1   ";
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6);
+
+  // 10 trials with random positions.
+  for (size_t i = 0; i < 10; ++i)
+  {
+    arma::mat coordinates(2, 2, arma::fill::randu);
+
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(LMNNFunctionGradientTest2)
+{
+  // Useful but simple dataset with six points and two classes.
+  arma::mat dataset        = "-0.1 -0.1 -0.1  0.1  0.1  0.1;"
+                             " 1.0  0.0 -1.0  1.0  0.0 -1.0 ";
+  arma::Row<size_t> labels = " 0    0    0    1    1    1   ";
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6);
+
+  // 10 trials with random positions.
+  for (size_t i = 0; i < 10; ++i)
+  {
+    arma::mat coordinates(2, 2, arma::fill::randn);
+
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(LMNNFunctionGradientTest3)
+{
+  arma::mat dataset;
+  arma::Row<size_t> labels;
+  data::Load("iris.csv", dataset);
+  data::Load("iris_labels.txt", labels);
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6);
+
+  // 10 trials with random positions.
+  for (size_t i = 0; i < 10; ++i)
+  {
+    arma::mat coordinates(2, 2, arma::fill::randn);
+
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(LMNNFunctionGradientTest4)
+{
+  arma::mat dataset;
+  arma::Row<size_t> labels;
+  data::Load("iris.csv", dataset);
+  data::Load("iris_labels.txt", labels);
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6);
+
+  // 10 trials with random positions.
+  for (size_t i = 0; i < 10; ++i)
+  {
+    arma::mat coordinates(2, 2, arma::fill::randu);
+
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
 BOOST_AUTO_TEST_SUITE_END();
```

(You could add that if you like.)

So it seems that nothing is wrong.  For the observations:

  1. This is expected, the step size for SGD has to be tuned.
  2. This may happen depending on the loss surface.
  3. This one concerns me a little bit more, but what is the dataset and parameters you are using?

I tried using the `mlpack_lmnn` program and added some bits that would compute the accuracy before and afterwards.  It is much faster when compiled without debugging symbols; the iris dataset with k=3 and SGD with default parameters takes about half a second on my machine.  But the dataset is fairly small, so usually the accuracy increase is pretty marginal (i.e., 96 -> 96.7%, etc.).

So I think it may be worth finding some larger datasets to try on, and seeing how it does there.  But it seems to me that the implementation is correct, based on the tests and experiments I did.  Maybe with other datasets it behaves differently---I am not sure.

manish7294 replied:
> This one concerns me a little bit more, but what is the dataset and parameters you are using?

There are not many configurations where we see this behavior and as per the observation by tuning the step_size in case of sgd we can get atleast the same accuracy.
Here are some examples:
1. iris -> k =2, default accuracy - 94.6667 , lbfgs accuracy - 94, sgd accuracy  - 94.6667 (this is the only case that exist in iris)
2. vc2 -> k = 4, default accuracy - 79.7101 , lbfgs accuracy - 77.7778, sgd accuracy  - 83.0918
3. vc2 -> k = 11, default accuracy - 82.1256 , lbfgs accuracy - 81.1594, sgd accuracy (after lot of tuning)  - 82.1256 
I think it may have to do with my accuracy calcuator https://pastebin.com/MFdVDcrz
Let's say we have k = 2 and we get two different points of two different classes, so it will always select the point with lower label value as the prediction.

rcurtin replied:
Ah, ok, I think you are right.  Technically the kNN rule should select the nearest neighbor label in case of a tie.

manish7294 replied:
Do you think, we should apply some weight parameter according to distance or just neighbor position here?
```
for (size_t j=0; j < k; j++)
      Map(labels(neighbors(j,i)))++;
```
example: instead of doing  Map(labels(neighbors(j,i)))++;
we can do `Map(labels(neighbors(j,i))) += 1 / j;` or something like else.

rcurtin replied:
Either is fine, both will solve the issue you mentioned.  I'm headed to bed now, I'll answer anything else in the morning.

manish7294 replied:
I think this is the exact problem, I tried Map(labels(neighbors(j,i)))+=std::exp(1/(j+1)) and the results were awesome.

manish7294 commented:
> Ah, by the way, are you configuring your build without debugging symbols and with optimization? (i.e. cmake -DDEBUG=OFF) That can give large speedups if you haven't done it already.

I have been using debugging symbols and some of the small optimizations that you have suggested, Big ones are still in the progress. Will follow this tip.

manish7294 commented:
I added a private member `iter` to the LMNNFunction class, and I am incrementing it by 1 at each Evaluate call. This change is working perfectly fine with small datasets like iris, vc2 but don't know why I keep on getting the segmentation fault with covertype.
Here is the change:
```
if (iter % 100 == 0)
{
  Constraints constraint(transformedDataset, labels, k);
  constraint.Impostors(impostors, begin, batchSize);
}
```
I have also removed precalculation as it was creating serious memory issues with covertype and wasn't much proficient in giving speedups.

manish7294 commented:
Passes vs KNN Accuracy Graphs (Starting point is initial accuracy on the original datasets in all the figures, Hyperparameters were -> k = 5, step_size = 0.001, Optimizer = Adagrad):

vc2 dataset:
![hpehqlu](https://user-images.githubusercontent.com/17105412/41334118-06ec5a8e-6f02-11e8-9aba-d7545ab9521a.png)

iris dataset:
![hpfcgx5](https://user-images.githubusercontent.com/17105412/41334169-3e42c928-6f02-11e8-9c8a-8539fa8bf610.png)

diabetes dataset:
![hpfosth](https://user-images.githubusercontent.com/17105412/41335499-9f991430-6f06-11e8-8c67-3d1e64004859.png)

manish7294 commented:
Benchmarks on Iris:
AmsGrad (passes - 100):
K = 5, Range - 20 -> Accuracy Initial: 96.0000, Accuracy Final: 96.6667, total_time: 0.864678s, computing_neighbors: 0.572140s
K = 10, Range - 20 -> Accuracy Initial: 96.0000, Accuracy Final: 96.6667, total_time: 1.451917s, computing_neighbors: 0.863207s

BBSGD (passes - 50, Range not working --- optimizer stucks after some point):
K = 5 -> Accuracy Initial: 96.0000, Accuracy Final: 96.0000, total_time: 14.260040s, computing_neighbors: 8.797722s
K = 10 -> Accuracy Initial: 96.0000, Accuracy Final: 96.6667, total_time: 24.585565s, computing_neighbors: 12.057315s

SGD (passes - 100):
K = 5, Range - 20 -> Accuracy Initial: 96.0000, Accuracy Final: 96.0000, total_time: 1.663173s, computing_neighbors: 1.118650s
K = 10, Range - 20 -> Accuracy Initial: 96.0000, Accuracy Final: 96.0000, total_time: 2.915806s, computing_neighbors: 1.774652s

L-BFGS:
K = 5, Range - 20 -> Accuracy Initial: 96.0000, Accuracy Final: 98.0000, total_time: 0.367947s, computing_neighbors: 0.069805s
K = 10, Range - 20 -> Accuracy Initial: 96.0000, Accuracy Final: 97.3333, total_time: 0.989437s, computing_neighbors: 0.109728s

Note:  Range stands for impostors calculation after N iterations.

manish7294 commented:
Benchmarks on vc2:
AmsGrad (passes - 100):
K = 5, Range - 20 -> Accuracy Initial: 78.2609, Accuracy Final: 81.6425, total_time: 3.208379s, computing_neighbors: 2.246410s
K = 10, Range - 20 -> Accuracy Initial: 81.6425, Accuracy Final: 82.1256, total_time: 4.481770s, computing_neighbors: 3.274698s

BBSGD (passes - 50, Range not working --- optimizer stucks after some point):
K = 5 -> Accuracy Initial: 78.2609, Accuracy Final: 83.5749, total_time: 36.541889s, computing_neighbors: 21.828721s
K = 10 -> Accuracy Initial: 81.6425, Accuracy Final: 83.0918, total_time: 64.257297s, computing_neighbors: 32.795085s

SGD (passes - 100):
K = 5, Range - 20 -> Accuracy Initial: 78.2609, Accuracy Final: 83.5749, total_time: 3.375260s, computing_neighbors: 2.426689s
K = 10, Range - 20 -> Accuracy Initial: 81.6425, Accuracy Final: 82.6087, total_time: 5.255458s, computing_neighbors: 3.199045s

L-BFGS:
K = 5, Range - 20 -> Accuracy Initial: 78.2609, Accuracy Final: 80.6763, total_time: 1.157363s, computing_neighbors: 0.171003s
K = 10, Range - 20 -> Accuracy Initial: 81.6425, Accuracy Final: 84.058, total_time: 3.869821s, computing_neighbors: 0.233824s

Note:  Range stands for impostors calculation after N iterations.

manish7294 commented:
> System : slake; Training Parameters : Optimizer : L-BFGS, max_iterations : 2000, Range : 20, k : 3

|   | mlpack | | | | | 
| ------------- | ------------- | ------------- |  ------------- | ------------- |  ------------- | 
|      |               Time        | Accuracy_3_NN | Accuracy_3_NN_DW | Accuracy_5_NN | Accuracy_5_NN_DW|
| iris |  0.028843  | 96.66666666666667 | 96.66666666666667 |97.33333333333334 |97.33333333333334 |
| satellite | 6.099969  | 89.58041958041957| 90.7459207459207 | 91.11888111888112 | 91.6083916083916 |
| satellite (Range: 1, iterations: 3000) | 23.672363  | 93.1002331002331| 93.16239316239316 | 94.14918414918415 | 94.07925407925408 |
| ecoli | 0.020087  |91.07142857142857 | 90.625 | 92.41071428571429 |92.85714285714286 |
| ecoli  (Range: 1, iterations: 3000) | 0.053707  |91.07142857142857 |  91.51785714285714 | 92.41071428571429 | 92.41071428571429 |
| vehicle | 0.620096  |78.84160756501181 | 78.95981087470449 |81.4420803782506 | 79.90543735224587 |
| balance | 0.071944 | 90.72 | 91.03999999999999 | 92.32000000000001 | 92.64 |
| letter | 19.975593  | 97.0  | 97.13000000000001 | 96.94500000000001 | 97.385 |

|   | shogun | | | | | 
| ------------- | ------------- | ------------- |  ------------- | ------------- |  ------------- | 
|      |               Time        | Accuracy_3_NN | Accuracy_3_NN_DW | Accuracy_5_NN | Accuracy_5_NN_DW|
| iris | 1.3402702808380127  | 97.33333333333334 | 97.33333333333334 |96.66666666666667 |96.66666666666667 |
| satellite | 122.91067862510681  | 94.77855477855478| 94.82517482517483 | 94.68531468531468 | 94.68531468531468 |
| ecoli | 5.9187331199646  |92.41071428571429 | 92.41071428571429 | 93.75 |93.75 |
| vehicle | 18.766937017440796  |80.02364066193853 | 77.89598108747045 |79.43262411347519 | 77.42316784869976 |
| balance | 10.840742826461792 |  81.28 | 80.47999999999999 | 84.48 | 84.96000000000001 |
| letter (force stop after running for about 1.5 hours) | -  | -  | - | - | - |

rcurtin commented:
These results look pretty great---in some cases though it seems like the Shogun implementation converges to much better results, like for the `ecoli` and `vehicle` datasets.  I suspect that we'd also see different results with SGD-like optimizers too.

I think we can get some additional speedup, but it is really nice to have somewhere to start from.  It could be useful to add the MATLAB LMNN code also, if you aren't already planning on doing that.  Thank you for taking the time to get all of the benchmarking system working!

When I find time I will be implementing the bound idea (should be later today or tomorrow?) and we can see what kinds of additional speedup that gives.  In any case we are definitely on a good track.

rcurtin commented:
I've spent some time playing around with the bounds that I derived (http://www.ratml.org/misc/lmnn_bounds.pdf) and I found that how useful these bounds are will depend significantly on the dataset.  If the norms of the points in the dataset are large, then the steps must be very small to guarantee no neighbors have changed.  Sometimes, mean centering may not help.  I found that with the covertype dataset, typically the average squared norm will be on the order of 1e7 to 1e8, and typical distance between nearest neighbors will be in the hundreds.  So for the bound to be satisfied, it would be necessary that the squared norm of (L_t - L_{t + 1}) would need to be something like 1e-6 or smaller, which typically does not happen until late in the LMNN learning process.  I found that when learning with L-BFGS with default parameters on the covertype-5k dataset, nearly every iteration has 3000+ (of 5000) points with changed impostors.

On the other hand, the iris dataset has points with small norms (mean-centered or not), so it turns out that only the first handful of iterations have changing impostors, and then they stay relatively constant.

So, my tentative conclusion is the following: using the bounds in the paper could be helpful, but not for all datasets.  Given the small cost of computing the dot product between the new and old coordinates, I think it would be worth doing.  Would you agree with that?

As far as other timing issues I see the following ideas:

 * Reducing the computation cost of calculating the impostors when it is necessary.  A lot of this time is spent building the trees and then performing the search.  So I think that we can save a lot of time if we can cache the trees that we build on the points, by building trees only once at the construction time of the LMNNFunction, using the `MahalanobisDistance` as a metric.  However, this will require the implementation of a bound like `MahalanobisHRectBound`, which will be similar to `HRectBound` (see `src/mlpack/core/tree/`), except that it will use an internal `L` matrix that can be changed.  

 * When we call Impostors() or TargetNeighbors(), we are computing the distance between a point and its impostors or target neighbors, but then in `EvaluateWithGradient()` we sometimes recalculate this.  I think at least for the impostors, we could cache these distance computations to save some time.

 * I bet we could save some computations by bounding this part:

```
        // Calculate cost due to {data point, target neighbors, impostors}
        // triplets.
        double eval = metric.Evaluate(transformedDataset.col(i),
                          transformedDataset.col(targetNeighbors(j, i))) -
                      metric.Evaluate(transformedDataset.col(i),
                          transformedDataset.col(impostors(l, i)));

        // Check bounding condition.
        if (eval < -1)
        { 
          // update bound.
          bp = l;
          break;
        }
```

Basically, we want to know if eval will be less than -1, and if so we can skip the computation.  But if eval is way less than -1, it's not likely to change the next iteration.  We can use the triangle inequality in the bounds to use the previous iteration's eval quantity and bound the maximum possible eval this iteration.  But in order to do that, it's probably worth investigating (a) how many points have `eval < -1` each iteration and (b) how that changes over the course of the optimization.

So I think there is still a lot of speedup left on the table!  The results you've got scale to the letters dataset (20k points?) but I'd like to be able to scale to more like 500k-1M points.  I am pretty sure it is possible!

rcurtin commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -155,12 +162,17 @@ double LMNNFunction<MetricType>::Evaluate(const arma::mat& coordinates,
-        // Calculate cost due to data point, target neighbors, impostors
+        // Check if we already have distance between impostor & data point
+        // stored.
+        double distImp = distance(l, i) > 0 ? distance(l, i) :
```
You could change the condition `distance(l, i) > 0` to `iteration % range == 0`, and then avoid the `arma::fill::zeros` in the constructor of the `distance` matrix.  Plus then we don't have to worry about the strange edge case when `distance(l, i)` is actually legitimately 0. :)

manish7294 replied:
Had exact same thoughts, but then proceeded with this one :) 
Will change it shortly. I also timed this approach and it look on iris we get a speed up of .1 sec

rcurtin replied:
Right, I think in many cases this could be a major speedup.  But I don't see any way to avoid the similar `targetNeighbors` distance computations.

rcurtin replied:
Actually are you sure the speedup is 0.1s?  In the table you showed iris only takes 0.02s.  If we made the computation take negative time I think we are definitely on to something good ;)

manish7294 replied:
Now I have to admit my system is no more than turtle, the second timing is on my system. And most probaby some other parameters :)

rcurtin replied:
Ah, ok, different systems, no worries :)

rcurtin commented on src/mlpack/methods/lmnn/lmnn_impl.hpp:
```diff
@@ -0,0 +1,58 @@
+      !(arma::is_finite(outputMatrix)))
+    outputMatrix.eye(dataset.n_rows, dataset.n_rows);
+
+  Timer::Start("lmnn_sgd_optimization");
```
This may or may not be `SGD`, so if we'll use a timer here I'd suggest just `lmnn_optimization`.

manish7294 replied:
Done!

rcurtin commented on src/mlpack/methods/lmnn/lmnn_main.cpp:
```diff
@@ -0,0 +1,350 @@
+    PRINT_PARAM_STRING("max_step") + " (which both refer to the line search "
+    "routine).  For more details on the L-BFGS optimizer, consult either the "
+    "mlpack L-BFGS documentation (in lbfgs.hpp) or the vast set of published "
+    "literature on L-BFGS."
```
This is great, thank you for having such clear and comprehensive documentation.  Could you add a bit about the `range`, `regularization`, and `normalization` parameters?  I think it would be nice to add an example call or two, do you think you could do that?  If you think that makes it too long, we can probably make the optimizer bits a little shorter, but I have no issue with long documentation.

manish7294 replied:
Done! Added examples too, Hope they are up to the expectations.

rcurtin commented on src/mlpack/methods/lmnn/lmnn_main.cpp:
```diff
@@ -0,0 +1,350 @@
+    "L-BFGS (0 indicates no limit).", "n", 100000);
+PARAM_DOUBLE_IN("tolerance", "Maximum tolerance for termination of AMSGrad, "
+    "BB_SGD, SGD or L-BFGS.", "t", 1e-7);
+PARAM_DOUBLE_IN("step_size", "Step size for AMSGrad, BB_SGD and SGD."
```
I think the `.` should be removed here and replaced with a space, so the `(alpha).` is included in the same sentence.

manish7294 replied:
Done!

manish7294 commented on src/mlpack/methods/lmnn/constraints_impl.hpp:
```diff
@@ -0,0 +1,319 @@
+
+    // Re-map neighbors to their index.
+    for (size_t j = 0; j < neighbors.n_elem; j++)
+      neighbors(j) = indexSame[i].at(neighbors(j));
```
I think we may not able to use transform as transform can't capture any other external variable----like indexSame[i] in this case.

rcurtin replied:
You're right, sorry---I guess that idea won't work here.

manish7294 commented on src/mlpack/methods/lmnn/constraints_impl.hpp:
```diff
@@ -0,0 +1,319 @@
+  Precalculate();
+
+  outputNeighbors = arma::Mat<size_t>(k, N, arma::fill::zeros);
+  outputDistance = arma::mat(k, N, arma::fill::zeros);
```
I have replaced them wherever possible :)

manish7294 commented on src/mlpack/methods/lmnn/lmnn.hpp:
```diff
@@ -0,0 +1,115 @@
+ * established by target neighbors should be large and that between target
+ * neighbors and data point should be small. It requires the knowledge of
+ * target neighbors beforehand. Moreover, target neighbors once initialized
+ * remain same.
```
Done!

manish7294 commented on src/mlpack/methods/lmnn/constraints_impl.hpp:
```diff
@@ -0,0 +1,319 @@
+  size_t N = dataset.n_cols;
+
+  // Perform pre-calculation, if necessary.
+  Precalculate();
```
Used the idea for both constraint and lmnn function.

manish7294 commented on src/mlpack/methods/lmnn/constraints_impl.hpp:
```diff
@@ -0,0 +1,319 @@
+
+  if (minCount < k)
+  {
+    Log::Fatal << "Constraints::VerifyK(): One of the class contains only "
```
Done!

manish7294 commented on src/mlpack/methods/lmnn/lmnn.hpp:
```diff
@@ -0,0 +1,115 @@
+ * target neighbors beforehand. Moreover, target neighbors once initialized
+ * remain same.
+ *
+ * @tparam SDPType The type of SDP to use for computation. ex. SDP<arma::mat>
```
Done!

manish7294 commented on src/mlpack/methods/lmnn/lmnn.hpp:
```diff
@@ -0,0 +1,115 @@
+   * both dataset and labels matrices to internal copies.
+   *
+   * @param dataset Input dataset.
+   * @param labels Input dataset labels.
```
Done!

manish7294 commented on src/mlpack/methods/lmnn/lmnn.hpp:
```diff
@@ -0,0 +1,115 @@
+  size_t& Range() { return objFunction.Range(); }
+
+  //! Access the value of k.
+  const size_t& K() const { return k; }
```
I am bit confused about doing this one. The main reason I left it was ---- On changing K we will have to modify all parts like recomputing target neighbors, impostors, making change in lmnn functions, etc. 
If you still want to make this change, then I will happily do that?

rcurtin replied:
Yeah, I think it would be useful.  Then `objFunction` could just be made local to the `LearnDistance()` function.  If you think it's better the way it is, I am okay with that too, but we should point out in the documentation (if it isn't mentioned already, I haven't checked) that K can only be changed by creating a new LMNN object.

manish7294 commented on src/mlpack/methods/lmnn/lmnn_function.hpp:
```diff
@@ -0,0 +1,205 @@
+   *
+   * @param covariance Covariance matrix of Mahalanobis distance.
+   */
+  double Evaluate(const arma::mat& covariance);
```
Changed to transformation matrix everywhere!

manish7294 commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,510 @@
+        {
+          eval = metric.Evaluate(transformedDataset.col(i),
+                    transformedDataset.col(targetNeighbors(j, i))) -
+                 std::pow(distance(l, i), 2);
```
Created a typedef for KNN

manish7294 commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,510 @@
+  transformedDataset = dataset;
+
+  // Initialize target neighbors & impostors.
+  Constraints constraint(dataset, labels, k);
```
Done!

manish7294 replied:
Ah! I added a modify dataset function to do the above. But somehow things broke up, had to revert back.

rcurtin replied:
Do you want to dig into this one a little more?  I think it's a worthwhile optimization to do before merge---it should be pretty simple.  I can help out if you like, just let me know.

manish7294 replied:
Done!

manish7294 commented on src/mlpack/methods/lmnn/constraints_impl.hpp:
```diff
@@ -0,0 +1,319 @@
+  {
+    // Perform KNN search with same class points as both reference
+    // set and query set.
+    knn.Train(dataset.cols(indexSame[i]));
```
We can do this by storing matrices in a vector of matrices but I wonder will it be a good thing to do.

rcurtin replied:
I think it's fine to do that.

manish7294 replied:
Ah! I think this won't work as dataset will keep on changing on every call, so it will be no good in storing it. Do you agree?

rcurtin replied:
I see what you mean.  I have an idea for how to do this; I'll open an issue.  No need to worry about it here though.

manish7294 commented:
@rcurtin  I have added Binding tests too.  As you get the chance, please have them reviewed as well.

manish7294 commented on src/mlpack/methods/lmnn/lmnn_function.hpp:
```diff
@@ -0,0 +1,205 @@
+  //! Range after which impostors need to be recalculated.
+  size_t range;
+  //! Holds pre-calculated cij.
+  arma::mat p_cij;
```
Done!

manish7294 commented on src/mlpack/methods/lmnn/lmnn_main.cpp:
```diff
@@ -0,0 +1,392 @@
+PARAM_MATRIX_IN_REQ("input", "Input dataset to run LMNN on.", "i");
+PARAM_UROW_IN("labels", "Labels for input dataset.", "l");
+PARAM_INT_IN("num_targets", "Number of target neighbors to use for each "
+    "datapoint.", "k", 1);
```
Done!

manish7294 commented on src/mlpack/tests/main_tests/lmnn_test.cpp:
```diff
@@ -0,0 +1,829 @@
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("labels", std::move(labels));
+  SetInputParam("optimizer",  std::string("bbsgd"));
+  SetInputParam("batch_delta", std::move(10.0));
```
Done!

manish7294 commented on src/mlpack/tests/main_tests/lmnn_test.cpp:
```diff
@@ -0,0 +1,829 @@
+ * Ensure that different value of regularization results in a
+ * different output matrix.
+ */
+BOOST_AUTO_TEST_CASE(LMNNDiffRegularizationTest)
```
Done!

manish7294 commented on src/mlpack/methods/lmnn/lmnn_main.cpp:
```diff
@@ -0,0 +1,392 @@
+PARAM_UROW_IN("labels", "Labels for input dataset.", "l");
+PARAM_INT_IN("num_targets", "Number of target neighbors to use for each "
+    "datapoint.", "k", 1);
+PARAM_MATRIX_OUT("output", "Output matrix for learned distance matrix.", "o");
```
Done!

manish7294 commented on src/mlpack/methods/lmnn/lmnn_main.cpp:
```diff
@@ -0,0 +1,392 @@
+    "the same will look like: "
+    "\n\n"
+    "$ bin/mlpack_lmnn -i iris.csv -l iris_labels.txt -k 3 -O bbsgd -o "
+    "output.csv --verbose"
```
Done!

manish7294 commented:
Ah! I just realized, there was no parameter for allowing the user to pass initial learning point, hence losing all significance of Low-Rank optimization. I have made some changes regarding it. Hopefully, they should suffice.

rcurtin commented on src/mlpack/tests/lmnn_test.cpp:
```diff
@@ -0,0 +1,551 @@
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
```
I think it would be a good idea to check that we can run LMNN with a low-rank distance matrix and we still get reasonable results.  You could create a high-dimensional random dataset (maybe two Gaussians or something, one for each class) and then run LMNN twice, once with a low-rank matrix and once with a normal matrix, and then make sure the KNN accuracies are comparable at the end of the two runs.  What do you think?

You could do that before merge or we can open a ticket for it, I am not picky.

manish7294 replied:
sure, I will do that.

manish7294 replied:
Done!

manish7294 commented on src/mlpack/methods/lmnn/lmnn_main.cpp:
```diff
@@ -0,0 +1,468 @@
+    1e20);
+PARAM_INT_IN("range", "Number of iterations after which impostors needs to be "
+    "recalculated", "R", 1);
+PARAM_INT_IN("seed", "Random seed.  If 0, 'std::time(NULL)' is used.", "s", 0);
```
Right, this eventually got so big, that I had a hard time finding alphabets for them :)
It would be good enough if we can just avoid all basic optimizers specific parameters, I guess there are many like beta1, beta2, etc. Probably they will be used occasionally.

rcurtin replied:
Ok---if you want to simplify before merge, feel free; otherwise we can open an issue.

manish7294 replied:
Done!

rcurtin replied:
I think it's fine to remove `num_basis`, `armijo_constant`, and `wolfe`, if you like, since these don't usually need to be tuned for L-BFGS either (in fact L-BFGS doesn't tend to need tuning much at all).

manish7294 commented on src/mlpack/methods/lmnn/lmnn_main.cpp:
```diff
@@ -0,0 +1,468 @@
+
+  if (CLI::HasParam("distance"))
+  {
+    distance = std::move(CLI::GetParam<arma::mat>("distance"));
```
That will be fair enough.

manish7294 replied:
Done!

manish7294 commented on src/mlpack/tests/main_tests/lmnn_test.cpp:
```diff
@@ -0,0 +1,936 @@
+ * Ensure that when we pass a valid initial learning point, we get
+ * output of the same dimensions.
+ */
+BOOST_AUTO_TEST_CASE(LMNNValidDistanceTest)
```
Ya, there is an LMNNInvalidDistanceTest as well just below it. I think that's what you're probably referring too.

manish7294 replied:
Sorry, If I not getting the point but I believe those are the exact same ideas valid and invalid distance tests are based on. Like valid one for 1st point and invalid one for 2nd point(checking square matrix). Please, correct me if I am off the point.

rcurtin replied:
Oh, sorry!  You are exactly right.  I had misread the valid distance test.  I thought that in this test, you were passing in a square matrix.

Do you think it would be worthwhile to then just add a third test to ensure that we can pass a distance matrix of size `inputData.n_rows` by `inputData.n_rows`?

manish7294 replied:
no worries! I will add it.

manish7294 replied:
Done!

manish7294 commented on src/mlpack/methods/lmnn/lmnn_main.cpp:
```diff
@@ -0,0 +1,468 @@
+
+    for (size_t j = 0; j < k; j++)
+      Map(labels(neighbors(j, i))) +=
+          1 / std::pow(distances(j, i) + 1, 2);
```
sure :)

zoq commented on src/mlpack/methods/lmnn/constraints.hpp:
```diff
@@ -0,0 +1,188 @@
+   */
+  Constraints(const arma::mat& dataset,
+              const arma::Row<size_t>& labels,
+              size_t k);
```
Should we use ``const`` here as well, to be consistent with the rest of the codebase?

zoq commented on src/mlpack/methods/lmnn/constraints.hpp:
```diff
@@ -0,0 +1,188 @@
+
+  /**
+   * Calculates k similar labeled nearest neighbors for a batch of dataset and
+   *  stores them into the passed matrix.
```
Looks like an extra space here.

zoq commented on src/mlpack/methods/lmnn/constraints.hpp:
```diff
@@ -0,0 +1,188 @@
+                const arma::mat& dataset,
+                const arma::Row<size_t>& labels);
+
+  //! Access the value of k.
```
Do you think we could improve the comment here? e.g. ``Get the number of target neighbors (k).``.

manish7294 replied:
Done!

zoq commented on src/mlpack/methods/lmnn/constraints_impl.hpp:
```diff
@@ -0,0 +1,328 @@
+
+    // Re-map neighbors to their index.
+    for (size_t j = 0; j < neighbors.n_elem; j++)
+      neighbors(j) = indexDiff[i].at(neighbors(j));
```
Just an idea, but could directly use ``outputMatrix`` and avoid the extra copy at the end?

manish7294 replied:
I think we can do that, but will be a bit messy.
```
// Re-map neighbors to their index.
for (size_t j = 0; j < neighbors.n_elem; j++)
  outputMatrix(j % k, indexSame[i](j / k)) = indexSame[i].at(neighbors(j));
```
Shall we do this?

zoq replied:
Your call, not sure we would see any performance improvements.

rcurtin commented on src/mlpack/tests/lmnn_test.cpp:
```diff
@@ -441,6 +441,55 @@ BOOST_AUTO_TEST_CASE(LMNNAccuracyTest)
+  // BOOST_REQUIRE_CLOSE(Accuracy1, Accuracy2, 2.0);
+}
+
 // Comprehensive gradient tests by Ryan Curtin.
```
Ha, technically, I think it was Marcus who originally wrote these and then I adapted them. :)

manish7294 replied:
Then both should be there :)

rcurtin commented on src/mlpack/tests/lmnn_test.cpp:
```diff
@@ -441,6 +441,55 @@ BOOST_AUTO_TEST_CASE(LMNNAccuracyTest)
+  double Accuracy2 = KnnAccuracy(outputMatrix * dataset, labels, 1);
+
+  // Keeping tolerance very high.
+  // BOOST_REQUIRE_CLOSE(Accuracy1, Accuracy2, 2.0);
```
I was thinking of checking something more like, e.g., check that both `Accuracy1` and `Accuracy2` are greater than some small value like 50%.  The concern is that if there is a problem with the low-rank LMNN, the accuracy would end up near 0.  It's fine to leave this commented out for now; let's discuss more in #1447.

ShikharJ commented:
It might also help to squash / fixup the commits into one? So that it's easier to review?

manish7294 commented:
@ShikharJ  Yeah that could simplify things but sorry, I can't do that as of now, at least not until Wednesday( I think merge date is somewhere between that period).  Sorry in case you are having any kind of unconvienience, please try to bear with it :)

zoq commented on src/mlpack/methods/lmnn/lmnn_function_impl.hpp:
```diff
@@ -0,0 +1,501 @@
+
+template<typename MetricType>
+LMNNFunction<MetricType>::LMNNFunction(const arma::mat& dataset,
+                           const arma::Row<size_t>& labels,
```
Another picky style comment from my side, do you mind to align the method parameter with the first one, if you like you can also put the first parameter on the next line.

zoq commented on src/mlpack/methods/lmnn/lmnn_function.hpp:
```diff
@@ -0,0 +1,210 @@
+   */
+  template<typename GradType>
+  double EvaluateWithGradient(const arma::mat& transformation,
+                            const size_t begin,
```
Can you check the method parameter alignment?

zoq commented on src/mlpack/methods/lmnn/lmnn_function.hpp:
```diff
@@ -0,0 +1,210 @@
+   * @tparam GradType The type of the gradient out-param.
+   * @param transformation Transformation matrix of Mahalanobis distance.
+   * @param begin Index of the initial point to use for objective function.
+   * @param batchSize Number of points to use for objective function.
```
This should come after the ``gradient`` parameter.

zoq commented on src/mlpack/methods/lmnn/lmnn_function.hpp:
```diff
@@ -0,0 +1,210 @@
+   * @tparam GradType The type of the gradient out-param.
+   * @param transformation Transformation matrix of Mahalanobis distance.
+   * @param begin Index of the initial point to use for objective function.
+   * @param batchSize Number of points to use for objective function.
```
See comment below.

zoq commented on src/mlpack/methods/lmnn/lmnn_function.hpp:
```diff
@@ -0,0 +1,210 @@
+  /**
+   * Constructor for LMNNFunction class.
+   *
+   * @param data Dataset for which metric is calculated.
```
Looks like the comments don't match with the actual parameter.

zoq commented on src/mlpack/methods/lmnn/lmnn_function.hpp:
```diff
@@ -0,0 +1,210 @@
+namespace mlpack {
+namespace lmnn {
+
+
```
Perhaps you can comment on what this class does?

manish7294 replied:
Done!

zoq commented on src/mlpack/methods/lmnn/constraints_impl.hpp:
```diff
@@ -0,0 +1,328 @@
+  for (size_t i = 0; i < uniqueLabels.n_elem; i++)
+  {
+    // Store same and diff indices.
+    indexSame[i] = arma::find(labels == uniqueLabels[i]);
```
I'm wondering if it makes sense to store the result of ``labels == uniqueLabels[i]`` so that it could be reused for the next line

manish7294 replied:
Doing so will need something like set substraction or an AND operation like `(labels && !(previousResult))`, which I think will do nothing good here, as it will be more or less the same thing.

rcurtin replied:
I think that as we solve #1445 this code will be removed, so I don't think we need to worry about it for now.

zoq commented on src/mlpack/methods/lmnn/constraints_impl.hpp:
```diff
@@ -0,0 +1,328 @@
+
+template<typename MetricType>
+inline void Constraints<MetricType>::Precalculate(
+                                        const arma::Row<size_t>& labels)
```
This should be tapped twice.

manish7294 replied:
Done!
